include/ensmallen_bits/pso/pso.hpp

```diff
@@ -0,0 +1,242 @@
+ * that is inspired by flocks or birds or fishes. The fundamental analogy is
+ * that every creature (particle in a swarm) is at a measurable position of
+ * `goodness' (in the context of PSO, called `fitness') in terms of being able
+ * to find food, and this information can be shared amongst the creatures in the
+ * flock, so that iteratively, the entire flock can get close to the optimal
+ * food source. In a more technical respect, the means by which the fitness
+ * information is shared determines the way in which the global optimum is
+ * approached.
+ *
+ * When this information is shared among particles whose fitness is close to
+ * each other (in a sense, the `nearest' neighbors in the fitness space), the
+ * variant of the approach is called the `local-best' or `lbest' PSO
+ * (consequently, it follows a ring-topology in an information-communication
+ * sense); and when this information is globally shared, the variant is called
+ * the `global-best' or `gbest' PSO (consequently, it follows a star-topology in
+ * an information-communication sense).
+ *
+ * For more information, refer to:
+ *
+ * @inproceedings{Kennedy,
+ *    doi = {10.1109/icnn.1995.488968},
+ *    url = {https://doi.org/10.1109/icnn.1995.488968},
+ *    publisher = {{IEEE}},
+ *    author = {J. Kennedy and R. Eberhart},
+ *    title = {Particle swarm optimization},
+ *    booktitle = {Proceedings of {ICNN}{\textquotesingle}95 - 
+ *                 International Conference on Neural Networks}
+ * }
+ *
+ * PSO can optimize arbitrary functions. For more details, see the documentation
+ * on function types included with this distribution or on the ensmallen
+ * website.
+ * 
+ * For PSO to work, the function being optimized must implement an
+ * ArbitraryFunctionType template parameter. The respective class must implement
+ * the following function:
+ *
+ *    double Evaluate(const arma::mat& x);
+ */
+template<typename VelocityUpdatePolicy = LBestUpdate,
+         typename InitPolicy = DefaultInit>
+class PSOType
+{
+ public:
+  /**
+   * Construct the particle swarm optimizer with the given function and
+   * parameters. The defaults here are not necessarily good for the given
+   * problem, so it is suggested that the values used be tailored to the task
+   * at hand.
+   *
+   * @param numParticles Number of particles in the swarm.
+   * @param lowerBound Lower bound of the coordinates of the initial population.
+   * @param upperBound Upper bound of the coordinates of the initial population.
+   * @param maxIterations Number of iterations allowed.
+   * @param horizonSize Size of the lookback-horizon for computing improvement.
+   * @param impTolerance Improvement threshold for termination.
+   * @param exploitationFactor Influence of the personal best of the particle.
+   * @param explorationFactor Influence of the neighbours of the particle.
+   */
+  PSOType(const size_t numParticles = 64,
+          arma::vec lowerBound = arma::ones<arma::vec>(1),
```

Comment:
Just noted, we should probably use ``arma::vec& lowerBound`` same for the ``upperBound`` parameter, sorry missed this.

Reply:
Yeah, that makes sense. I'm sorry for missing this, too.

Reply:
There's one problem - if we use `arma::vec& lowerBound`, then we cannot use a default value for `lowerBound` and `upperBound`.

Reply:
``const arma::vec& lowerBound = arma::ones<arma::vec>(1)`` should work.

Reply:
Thanks for the pointers. I've done this!